hnix-store-core/tests/Nar.hs

```diff
@@ -0,0 +1,150 @@
+{-# LANGUAGE OverloadedStrings #-}
```

Comment:
Hmm I feel like nar serialization/deserialization is a perfect case for property testing (quickcheck or whatever). Would you mind giving that a spin?

Reply:
Maybe in addition to tests derived from seeing what the nix cli does.

Reply:
Actually maybe we can just for now assume that the CLI is available and run it during the tests, instead of pre-generating fixtures.

Reply:
I'll add an `Arbitrary` sized instance and some tests that shell out to `nix-store`. Will keep the hand-copied fixtures for when "we win" and there is no more reference `nix-store` to use for tests :P

Reply:
Actually, I'm stumped about how you could _ever_ encode a string's length, followed by the string itself, without forcing the full string....

But looking at the source for `ByteString`, things look promising! Strict bytestrings give `length` in O(1), and it doesn't look like it forces the underlying bytestring. https://hackage.haskell.org/package/bytestring-0.10.8.2/docs/src/Data.ByteString.html#length ... and in the lazy case, `BSL.length bsl` is defined with a foldl' over the chunks ( https://hackage.haskell.org/package/bytestring-0.10.8.2/docs/src/Data.ByteString.Lazy.html#length ), which is strict in the `Int64` being accumulated, but shouldn't force the whole of `bsl`.... I'm not really sure I get the laziness going on here correctly, need to do an experiment.

Reply:
@imalsogreg You can do it by trusting the filesystem to correctly report the size of the file being read, and hoping the file doesn't change in the mean time (if it does, then you need to just error out in a streaming context)